optuna/storages/_rdb/storage.py

```diff
@@ -802,28 +811,21 @@ def _get_trials(
                 assert isinstance(states, Iterable)
                 query = query.filter(models.TrialModel.state.in_(states))
 
-            trial_ids = query.all()
-
-            trial_ids = set(
-                trial_id_tuple[0]
-                for trial_id_tuple in trial_ids
-                if trial_id_tuple[0] not in excluded_trial_ids
-            )
             try:
-                trial_models = (
-                    session.query(models.TrialModel)
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.params))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.values))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.user_attributes))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.system_attributes))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.intermediate_values))
-                    .filter(
-                        models.TrialModel.trial_id.in_(trial_ids),
-                        models.TrialModel.study_id == study_id,
+                if len(included_trial_ids) > 0 and trial_id_greater_than > -1:
+                    _query = query.filter(
+                        sqlalchemy.or_(
+                            models.TrialModel.trial_id.in_(included_trial_ids),
+                            models.TrialModel.trial_id > trial_id_greater_than,
+                        )
                     )
-                    .order_by(models.TrialModel.trial_id)
-                    .all()
-                )
+                elif trial_id_greater_than > -1:
+                    _query = query.filter(models.TrialModel.trial_id > trial_id_greater_than)
+                elif len(included_trial_ids) > 0:
+                    _query = query.filter(models.TrialModel.trial_id.in_(included_trial_ids))
```

Comment:
In my understanding, `trial_id_greater_than == -1` means that no trials are cached. So, I think that this filter is incorrect and no filter is appropriate.

Reply:
Thank you so much for pointing it out! So the code should look like below, right?

```python
if len(included_trial_ids) > 0 and trial_id_greater_than > -1:
    query = query.filter(sqlalchemy.or_(...))
elif trial_id_greater_than > -1:
    query = query.filter(models.TrialModel.trial_id > trial_id_greater_than)
```

Reply:
Yes, that's right. Additionally, we can sometimes simplify the filter by the following preprocessing before this block.

```python
included_trial_ids = [trial_id for trial_id in included_trial_ids if trial_id <= trial_id_greater_than]
```

Reply:
Thanks for the comment.

> Additionally, we can sometimes simplify the filter by the following preprocessing before this block.
Could you tell me how the code can be simplified by this preprocessing?

Reply:
Please note that it simplifies the filter, not the code.

----

For example, Observing `included_trial_ids` and `trial_id_greater_than` with the following code, we see the following behavior.

```python
import optuna

def objective(trial):
    return trial.suggest_float("x", 0, 1)

study = optuna.create_study(storage="sqlite:///tmp.db")
study.optimize(objective, n_trials=3)
```

```
[I 2024-10-24 14:09:07,513] A new study created in RDB with name: no-name-4641befe-65be-4946-b37a-d6233194fed0
included_trial_ids=set() trial_id_greater_than=-1
included_trial_ids={1} trial_id_greater_than=-1
[I 2024-10-24 14:09:07,571] Trial 0 finished with value: 0.8963270182225628 and parameters: {'x': 0.8963270182225628}. Best is trial 0 with value: 0.8963270182225628.
included_trial_ids={1} trial_id_greater_than=-1
included_trial_ids={2} trial_id_greater_than=1
[I 2024-10-24 14:09:07,605] Trial 1 finished with value: 0.7990099671251019 and parameters: {'x': 0.7990099671251019}. Best is trial 1 with value: 0.7990099671251019.
included_trial_ids={2} trial_id_greater_than=1
included_trial_ids={3} trial_id_greater_than=2
[I 2024-10-24 14:09:07,638] Trial 2 finished with value: 0.008839718644538253 and parameters: {'x': 0.008839718644538253}. Best is trial 2 with value: 0.008839718644538253.
```

With the suggested preprocessing, we can see that there are no redundant `included_trial_ids`.

```
[I 2024-10-24 14:09:17,389] A new study created in RDB with name: no-name-87d7c55f-0a10-4e59-9068-12d2b68e05ba
included_trial_ids=[] trial_id_greater_than=-1
included_trial_ids=[] trial_id_greater_than=-1
[I 2024-10-24 14:09:17,438] Trial 0 finished with value: 0.1040626734970852 and parameters: {'x': 0.1040626734970852}. Best is trial 0 with value: 0.1040626734970852.
included_trial_ids=[] trial_id_greater_than=-1
included_trial_ids=[] trial_id_greater_than=1
[I 2024-10-24 14:09:17,467] Trial 1 finished with value: 0.5720482628068786 and parameters: {'x': 0.5720482628068786}. Best is trial 0 with value: 0.1040626734970852.
included_trial_ids=[] trial_id_greater_than=1
included_trial_ids=[] trial_id_greater_than=2
[I 2024-10-24 14:09:17,498] Trial 2 finished with value: 0.410735588216247 and parameters: {'x': 0.410735588216247}. Best is trial 0 with value: 0.1040626734970852.
```

Reply:
I see. I think we don't have to use this preprocessing here since the actual SQL query is expanded to use `WHERE IN` clauses by SQLAlchemy, so speed up wouldn't be realized.

Reply:
I think the preprocessing also changes the SQL query. Please check the following example.

```
$ git diff
diff --git a/optuna/storages/_rdb/storage.py b/optuna/storages/_rdb/storage.py
index 85f47195f..db4ac3fc3 100644
--- a/optuna/storages/_rdb/storage.py
+++ b/optuna/storages/_rdb/storage.py
@@ -790,6 +790,8 @@ class RDBStorage(BaseStorage, BaseHeartbeat):
         included_trial_ids: Set[int],
         trial_id_greater_than: int,
     ) -> List[FrozenTrial]:
+        included_trial_ids = [trial_id for trial_id in included_trial_ids if trial_id <= trial_id_greater_than]
+
         with _create_scoped_session(self.scoped_session) as session:
             # Ensure that the study exists.
             models.StudyModel.find_or_raise_by_id(study_id, session)
```

```python
import logging
import optuna

def objective(trial):
    return trial.suggest_float("x", 0, 1)

logging.basicConfig()
logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)
study = optuna.create_study(storage="sqlite:///tmp.db")
study.optimize(objective, n_trials=3)
```

- before

```
INFO:sqlalchemy.engine.Engine:SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete 
FROM trials 
WHERE trials.study_id = ? AND (trials.trial_id IN (?) OR trials.trial_id > ?) ORDER BY trials.trial_id
```

- after

```
INFO:sqlalchemy.engine.Engine:SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete 
FROM trials 
WHERE trials.study_id = ? AND trials.trial_id > ? ORDER BY trials.trial_id
```

Reply:
Sounds very good. Thank you for your explanation. I applied your change.